### PR TITLE
[WIP] Reordering hidden columns

### PR DESCRIPTION
**Pull request by vitch**
This PR addresses #152 by tweaking the way reordering is done...

From the commit message of the last commit:

> In order to make the previously added test pass we need to give
the column reordering plugin access to _all_ columns, not just the
visible ones.
> When moving left or right we need to iteratively swap the position
of adjacent columns, continuing until we have swapped with a visible
column.
> There are some failing tests for internals of the component on this
commit but they can be fixed up later. I want to see if the new UX
is desirable.

There is also some implementation tweaks I'd like to make (using a
simple `for` loop would be more efficient because I could use the index
in the array to do the swapping) but I can do that in a separate
commit. 

I'm also not sure if we could do away with or simplify  the entire 
`swapWith` function (and maybe even `orderOf`?). There are some 
`TODO`s in the code as well for edge cases to clean up. At the moment
I'm looking for verification that the approach seems sound and the 
generated UX is an improvement...

### CODE REVIEW

File: test-app/tests/plugins/column-reordering/rendering-test.gts [hunk 1]
```diff
@@ -297,11 +297,12 @@ module('Plugins | columnReordering', function (hooks) {
       assert.strictEqual(getColumnOrder(), 'A B C D');
     });
 
-    test('without setting the order of anything, we cannot retain the order of the columns when they are added or removed', async function (assert) {
+    test.skip('without setting the order of anything, we retain the order of the columns when they are added or removed', async function (assert) {
```
Comment (vitch) [inline test-app/tests/plugins/column-reordering/rendering-test.gts:300]: This test is currently skipping because it fails. 

The plugin no longer supports the columns being mutated after the initial render from outside the table. Is this a legitimate thing that consumers should be able to do or was it just a simple way to initially write this test?

cc @NullVoxPopuli

Comment (NullVoxPopuli) [reply test-app/tests/plugins/column-reordering/rendering-test.gts:300]: Maybe, there are 3 use cases that i'm aware of:
- reorder each column live / immediately (like in this repos demos)
- reorder after submitting a form (setting the whole order at once) -- this is something common in table settings menus -- and relie(d|s) on a duplicate version of the ColumnOrder class (i forget the name) to have the same behavior as the columns.
- same as above, but without form submit.editing in a menu immediately changes the table (a hybrid of the prior two approaches, ux wise)

In all 3 of these, order is mutated from outside the table, so maybe i'm missing something, and need to go through the code.

Comment (vitch) [reply test-app/tests/plugins/column-reordering/rendering-test.gts:300]: > reorder each column live / immediately (like in this repos demos)
This works by using the plugins API, not by mutating the passed array

> reorder after submitting a form
There is an API for this too (`setAll` IIRC)

> editing in a menu immediately changes the table
I think that could use the API too...

I guess the bigger question is - should we expect consumers to use provided APIs to do things or should we let them mutate the inputs?

Comment (NullVoxPopuli) [reply test-app/tests/plugins/column-reordering/rendering-test.gts:300]: > This works by using the plugins API, not by mutating the passed array

Ah i see, that was the key part i was missing when you said mutation -- like a TrackedArray that has had additional columns added or removed or something.

> should we expect consumers to use provided APIs to do things or should we let them mutate the inputs?

I mean, one of the nice things deriving data from inputs gets us is that all of that can be handled.

Like, it could be a choice to require tearing down and recreating whatever when additional data is added to the table, but predefining all that before rendering is a common enough pattern, where it should probably be ok... If people want to asynchronously add more columns, they could probably deal with extra computation needed to re-position, re-size, etc the existing columns in addition to the new columns.

Like, personally, i'm not a fan of (what i see as) breaking reactivity (we can't know how deep or shallow a reactive abstraction is, and when we assume any depth, we can inherently any consuming usage -- now, maybe the maintainence burden in too great here, and that's fine! An assertion or something could be added to communicate that mutating the array passed to the ColumnOrder class isn't supported (tho, idk how common of a use case it is -- it's all kind of 'possibilities' here ))

**Comment by vitch** [comment]
> @vitch hey, great work.
> 
> I'm curious: What did you end up doing here – are you using this functionality in a fork?

I'm afraid I did the work because I needed it at work and I no longer work there... @joelamb might know if there is any active work on it?

**Comment by NullVoxPopuli** [comment]
Would there be interest in more distributed maintenance? 

repo could be transferred to https://github.com/universal-ember/ ?
